File: hackerrank/problem solving/algorithms/2 - implementation/30 - extra long factorials/extra_long_factorials.rs
// https://www.hackerrank.com/challenges/extra-long-factorials/problem?isFullScreen=true

use num::BigUint;

fn main() {
    let n: u32 = text_io::read!();
    println!("{}", factorial(n));
}

// T: O(n)
// S: O(1) extra space (excluding the size of the resulting integer)
fn factorial(n: u32) -> BigUint {
    (1..=n).product()
}

File: hackerrank/problem solving/algorithms/2 - implementation/30 - extra long factorials/extra_long_factorials.ts
// https://www.hackerrank.com/challenges/extra-long-factorials/problem?isFullScreen=true

function main() {
  const n: number = +readLine();
  console.log(factorial(n).toString());
}

// T: O(n)
// S: O(1) extra space (excluding the size of the resulting integer)
function factorial(n: number): BigInt {
  let res = BigInt(1);
  for (let i = BigInt(1); i <= n; i++) res *= i;
  return res;
}

//////////////////////////////////////////////////

import readline = require('readline');

const rl = readline.createInterface({ input: process.stdin });
let inputLines: string[] = [];
rl.on('line', (line: string) => inputLines.push(line));
rl.on('close', main);

function readLine(): string {
  return inputLines.shift()!;
}
